### PR TITLE
SCC: add AllowedFlexVolumes to manage a whitelist of allowed flexvolumes drivers

### PR DESCRIPTION
**Pull request by php-coder**
Proposal: https://github.com/kubernetes/community/pull/723
Trello: https://trello.com/c/YT6sNEay/61-5-sccfsi-psp-scc-flex-volume-support
Examples: https://github.com/openshift/origin/pull/15558#issuecomment-319702298

### CODE REVIEW

**Comment by deads2k** [comment]
@stevekuznetsov is there an easy way to trigger auto-labeling when a types.go file is modified?

File: pkg/security/apis/security/v1/types.go [hunk 1]
```diff
@@ -81,6 +81,9 @@ type SecurityContextConstraints struct {
 	// used to generate a value for a pod the first non-wildcard profile will be used as
 	// the default.
 	SeccompProfiles []string `json:"seccompProfiles,omitempty" protobuf:"bytes,20,opt,name=seccompProfiles"`
+	// AllowedFlexDrivers is a whitelist of allowed Flexvolume drivers.
+	// Empty or nil indicates that all drivers may be used.
+	AllowedFlexDrivers []string `json:"allowedFlexDrivers" protobuf:"bytes,21,opt,name=allowedFlexDrivers"`
```
Comment (deads2k) [inline pkg/security/apis/security/v1/types.go]: So this field is used when the user already has flexVolume power right?  If they don't, then this is never used?

If that's the case, I'd expect a validation rule and some more detail in the doc.  Also, why not move up underneath `Volumes`?  The proto number can still be 21 up there.

File: pkg/security/securitycontextconstraints/provider.go [hunk 2]
```diff
@@ -292,19 +292,38 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
...
+						fmt.Sprintf("%s volumes are not allowed to be used", string(fsType))))
+				}
 			}
+		}
 
-			if !allowedVolumes.Has(string(fsType)) {
-				allErrs = append(allErrs, field.Invalid(
-					fldPath.Child("volumes").Index(i), string(fsType),
-					fmt.Sprintf("%s volumes are not allowed to be used", string(fsType))))
+		if len(s.scc.AllowedFlexDrivers) > 0 && sccutil.SCCAllowsFSType(s.scc, securityapi.FSTypeFlexVolume) {
```
Comment (deads2k) [inline pkg/security/securitycontextconstraints/provider.go]: if the SCC doesn't allow flex volumes and you have any, wouldn't have already put errors on the return?

File: pkg/security/securitycontextconstraints/provider.go [hunk 3]
```diff
@@ -292,19 +292,38 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
...
+			for i, v := range pod.Spec.Volumes {
+				if v.FlexVolume == nil {
+					continue
+				}
+
+				for _, allowedDriver := range s.scc.AllowedFlexDrivers {
+					driver := v.FlexVolume.Driver
+					if driver != allowedDriver {
```
Comment (deads2k) [inline pkg/security/securitycontextconstraints/provider.go]: doesn't this mean that if I allow `["foo", "bar"]` and request one with `"bar"`, I'll get an error even though I'm allowed?  How about adding some tests here.

**Comment by stevekuznetsov** [comment]
@deads2k yes IIRC

**Comment by kargakis** [comment]
> @stevekuznetsov is there an easy way to trigger auto-labeling when a types.go file is modified?

Yes, there is. Will work on it today or tomorrow.

File: pkg/security/securitycontextconstraints/provider.go (hunk 3 above)
Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: Good catch, thanks! It's fixed now. I'll add tests after passing API review.

File: pkg/security/apis/security/types.go [hunk 4]
```diff
@@ -38,6 +38,10 @@ type SecurityContextConstraints struct {
 	// You must not list a capability in both AllowedCapabilities and RequiredDropCapabilities.
 	// To allow all capabilities you may use '*'.
 	AllowedCapabilities []kapi.Capability
+	// AllowedFlexDrivers is a whitelist of allowed Flexvolume drivers.
+	// Empty or nil indicates that all drivers may be used.
+	// +optional
+	AllowedFlexDrivers []string
```
Comment (liggitt) [inline pkg/security/apis/security/types.go]: making this an array of structs lets us expand options in the future (like setting the readOnly option), similar to the discussion around allowedHostPaths.

File: pkg/security/apis/security/v1/types.go (hunk 1 above)
Comment (php-coder) [inline pkg/security/apis/security/v1/types.go]: > So this field is used when the user already has flexVolume power right? If they don't, then this is never used?

Correct.

> If that's the case, I'd expect a validation rule and some more detail in the doc.

At present, it's possible to have non-empty `AllowedFlexDrivers` and prohibited the usage of flexVolumes at the same time. Should we add validation to that case, and force user to have an empty `AllowedFlexDrivers`?

> Also, why not move up underneath Volumes? The proto number can still be 21 up there.

Hm. It's not clear what do you mean, so I put `AllowedFlexDrivers` field right after `Volumes` in the `pkg/security/apis/security/types.go` file. I left the order the same in `pkg/security/apis/security/v1/types.go` file, let me know if you want to see it updated too.

Comment (php-coder) [inline pkg/security/apis/security/v1/types.go]: > Also, why not move up underneath Volumes?

Got it. Done.

File: pkg/security/securitycontextconstraints/provider.go (hunk 2 above)
Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: I'm sorry but I didn't understand what you mean here :(

Comment (liggitt) [inline pkg/security/securitycontextconstraints/provider.go]: if the volume type is completely disallowed, line 306 would already have added an error. there's no need to add another error saying the specific flex driver is not allowed. making a single loop over all the volumes would let you short-circuit adding specific errors if there are general ones.

```
allowAllVolumeTypes := sccutil.SCCAllowsAllVolumes(s.scc)
allowedVolumeTypes := sccutil.FSTypeToStringSet(s.scc.Volumes)
for i, v := range pod.Spec.Volumes {
  fsType, err := sccutil.GetVolumeFSType(v)
  if err != nil {
    ... add the error...
    continue
  }

  if !allowAllVolumeTypes && !allowedVolumes.Has(string(fsType)) {
    ... add an error about this volume type not being allowed...
    continue
  }

  if v.FlexVolume != nil && len(s.scc.AllowedFlexDrivers) > 0 {
    ... ensure the particular flex volume is allowed...
  }
}
```

File: pkg/oc/cli/describe/describer.go [hunk 5]
```diff
@@ -1865,6 +1866,13 @@ func stringOrNone(s string) string {
 	return "<none>"
```
Comment (pweil-) [inline pkg/oc/cli/describe/describer.go]: nit: change this to call `stringOrDefaultValue`

Comment (php-coder) [inline pkg/oc/cli/describe/describer.go]: Fixed.

File: pkg/security/apis/security/types.go (hunk 4 above)
Comment (php-coder) [inline pkg/security/apis/security/types.go]: Fixed.

**Comment by php-coder** [comment]
How to fix the following error?
```
++ Building go targets for linux/amd64: cmd/openshift cmd/oc cmd/kubefed pkg/sdn/plugin/sdn-cni-plugin vendor/github.com/containernetworking/cni/plugins/ipam/host-local vendor/github.com/containernetworking/cni/plugins/main/loopback
# github.com/openshift/origin/pkg/security/apis/security/v1
pkg/security/apis/security/v1/generated.pb.go:3450: undefined: allowedFlexvolume
```
I tried to run `hack/update-generated-protobuf.sh` many times but it didn't help.

**Comment by php-coder** [comment]
> How to fix the following error?

```
<claytonc> semushin: your "casttype=allowedFlexVolumes" is the problem
<claytonc> did you set the protobuf tags yourself?
<claytonc> you should remove the protobuf tags from your new generated code
<claytonc> remove the changes to generated.pb.go
<claytonc> then regenerate
<semushin> claytonc, yes, I added them manually (do we have another way?)
<claytonc> semushin: they are generated for you automatically by generated-protobuf
<claytonc> never add protobuf tags by hand unless you have a reason to (the reasons are small)
<claytonc> only add json tags
<claytonc> and honestly we should have json tag generation but it's easier today
```

**Comment by php-coder** [comment]
Test flake #10327
/test cmd

**Comment by php-coder** [comment]
Examples:

* `oc describe` when all allowed:
  ```console
  $ oc describe scc restricted | grep Flex
    Allowed Flexvolumes:			<all>
  ```
* `oc describe` with 2 allowed volumes:
  ```console
  $ oc describe scc restricted | grep Flex
    Allowed Flexvolumes:			driver=google/plus,driver=foo
  ```
* `oc get -o yaml`:
  ```console
  $ oc get scc restricted -o yaml  | grep Flex
  allowedFlexVolumes: []
  ```
* error message when pod request fails:
  > Error from server (Forbidden): error when creating "testing-flexvolume.json": pods "test-flex" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.containers[0].securityContext.volumes[1]: Invalid value: "example/bash": Flexvolume driver is not allowed to be used]
* `oc edit` fails when flex volumes aren't allowed to be used but the `AllowedFlexVolumes` field isn't empty:
  ```
  # securitycontextconstraints "restricted" was not valid:
  # * allowedFlexVolumes: Invalid value: []security.AllowedFlexVolume{security.AllowedFlexVolume{Driver:""}, security.AllowedFlexVolume{Driver:""}}: if usage of Flexvolumes are disallowed, the list of allowed Flexvolumes must be empty
  ```
* `oc edit` fails when `driver` wasn't specified:
  ```
  # securitycontextconstraints "restricted" was not valid:
  # * allowedFlexVolumes[0].driver: Required value: must specify a driver
  ```

**Comment by php-coder** [comment]
@openshift/api-review PTAL

**Comment by php-coder** [comment]
@openshift/api-review PTAL

**Comment by smarterclayton** [comment]
API lgtm

**Comment by php-coder** [comment]
Tests were added and code has been updated. PTAL.

**Comment by php-coder** [comment]
@openshift/security PTAL

File: pkg/security/apis/security/types.go [hunk 6]
```diff
@@ -42,6 +42,11 @@ type SecurityContextConstraints struct {
 	// of a VolumeSource (azureFile, configMap, emptyDir).  To allow all volumes you may use "*".
 	// To allow no volumes, set to ["none"].
 	Volumes []FSType
+	// AllowedFlexVolumes is a whitelist of allowed Flexvolumes.  Empty or nil indicates that all
```
Comment (wanghaoran1988) [inline pkg/security/apis/security/types.go]: s/Flexvolumes/FlexVolumes

**Comment by wanghaoran1988** [comment]
I see you use "Flexvolume" instead of "FlexVolume", would you intend to do this ?

File: pkg/security/apis/security/types.go (hunk 6 above)
Comment (php-coder) [inline pkg/security/apis/security/types.go]: Here I'm talking about the volumes of type flex and not about structs. [In the proposal about Flexvolumes](https://github.com/kubernetes/community/blob/8eca356fcb06d336f5128d93a9f68ef3f3867286/contributors/devel/flexvolume.md) they called as "Flexvolumes" and I'm using the same name here.

**Comment by php-coder** [comment]
@deads2k @enj PTAL

**Comment by php-coder** [comment]
Test flake #15874
/test extended_conformance_install_update

**Comment by php-coder** [comment]
Test flake #15874 again

**Comment by php-coder** [comment]
[(5) [SCCFSI] PSP/SCC Flex volume support](https://trello.com/c/YT6sNEay/61-5-sccfsi-psp-scc-flex-volume-support)

**Comment by php-coder** [comment]
`extended_conformance_gce` fails with:
```
[INFO] [CLEANUP] Pruning etcd data directory
rm: cannot remove ‘/tmp/etcd’: Operation not permitted
[ERROR] test/extended/conformance.sh exited with code 1 after 00h 30m 10s
make: *** [test-extended] Error 1
```
@stevekuznetsov Is it known bug or it's not even a bug? Here is a full log: https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_conformance_gce/6326/consoleFull

**Comment by smarterclayton** [comment]
Fixing latest flake in https://github.com/openshift/origin/pull/15904

**Comment by stevekuznetsov** [comment]
@php-coder that failure is from a best-effort bit of code so it's not killing your job. This is:

```
Summarizing 1 Failure:

[Fail] deploymentconfigs when run iteratively [Conformance] [It] should only deploy the last deployment 
/tmp/openshift/build-rpm-release/tito/rpmbuild-originFxMtUq/BUILD/origin-3.7.0/_output/local/go/src/github.com/openshift/origin/test/extended/deployments/deployments.go:84

Ran 206 of 815 Specs in 812.075 seconds
FAIL! -- 205 Passed | 1 Failed | 4 Pending | 605 Skipped 
```

File: pkg/security/apis/security/validation/validation.go [hunk 7]
```diff
@@ -77,17 +77,36 @@ func ValidateSecurityContextConstraints(scc *securityapi.SecurityContextConstrai
 			"required capabilities must be empty when all capabilities are allowed by a wildcard"))
 	}
 
-	if len(scc.Volumes) > 1 {
-		hasNone := false
+	hasFlexVolume := false
```
Comment (liggitt) [inline pkg/security/apis/security/validation/validation.go]: s/hasFlexVolume/allowsFlexVolumes/

File: pkg/security/apis/security/validation/validation.go [hunk 8]
```diff
@@ -77,17 +77,36 @@ func ValidateSecurityContextConstraints(scc *securityapi.SecurityContextConstrai
...
+		if hasFlexVolume {
+			for idx, allowedFlexVolume := range scc.AllowedFlexVolumes {
+				if len(allowedFlexVolume.Driver) == 0 {
+					allErrs = append(allErrs, field.Required(field.NewPath("allowedFlexVolumes").Index(idx).Child("driver"),
+						"must specify a driver"))
+				}
+			}
+		} else {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("allowedFlexVolumes"), scc.AllowedFlexVolumes,
+				"if usage of Flexvolumes are disallowed, the list of allowed Flexvolumes must be empty"))
```
Comment (liggitt) [inline pkg/security/apis/security/validation/validation.go]: reference the value that must be in the volumes list, e.g. `volumes does not include 'flexVolume' or '*', so no flex volumes are allowed`

File: pkg/security/securitycontextconstraints/provider.go [hunk 9]
```diff
@@ -269,6 +269,28 @@ func (s *simpleProvider) ValidatePodSecurityContext(pod *api.Pod, fldPath *field
 
 	allErrs = append(allErrs, s.sysctlsStrategy.Validate(pod)...)
 
+	if len(pod.Spec.Volumes) > 0 && len(s.scc.AllowedFlexVolumes) > 0 && sccutil.SCCAllowsFSType(s.scc, securityapi.FSTypeFlexVolume) {
```
Comment (liggitt) [inline pkg/security/securitycontextconstraints/provider.go]: odd... I expected this to extend the existing volume validation. should that move from ValidateContainerSecurityContext here as well?

File: pkg/security/apis/security/validation/validation.go [hunk 10]
```diff
@@ -77,17 +77,36 @@ func ValidateSecurityContextConstraints(scc *securityapi.SecurityContextConstrai
...
+
+	if hasNoneVolume && len(scc.Volumes) > 1 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("volumes"), scc.Volumes,
+			"if 'none' is specified, no other values are allowed"))
+	}
+
+	if len(scc.AllowedFlexVolumes) > 0 {
+		if hasFlexVolume {
+			for idx, allowedFlexVolume := range scc.AllowedFlexVolumes {
+				if len(allowedFlexVolume.Driver) == 0 {
```
Comment (wanghaoran1988) [inline pkg/security/apis/security/validation/validation.go]: I guess maybe you need validate the driver name format(we should have "openshift.com/foo" like driver name), if "foo" is specified instead of "openshift.io/foo", we should not pass the validation.

Comment (liggitt) [inline pkg/security/apis/security/validation/validation.go]: we shouldn't be stricter than the flex volume spec itself, and that does not validate the driver name at all

File: pkg/security/securitycontextconstraints/provider.go (hunk 9 above)
Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: Yes, it should. Kubernetes done it already: https://github.com/kubernetes/kubernetes/pull/31927

Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: P.S. But it should be in a separate PR.

Comment (liggitt) [inline pkg/security/securitycontextconstraints/provider.go]: if you want to move it separately, that's ok, but this PR should add to the existing volumes validation block

Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: Why? These checks are different and having them in a different places for some time shouldn't break anything or cause bugs. Or what I'm missing?

Comment (liggitt) [inline pkg/security/securitycontextconstraints/provider.go]: keeping the volume-related code together makes it easier to read and easier to refactor in the future

Comment (php-coder) [inline pkg/security/securitycontextconstraints/provider.go]: I don't want to modify this PR, could we merge it when tests will be green and postpone this refactoring a bit?

Comment (liggitt) [inline pkg/security/securitycontextconstraints/provider.go]: I'd prefer the volume validation stay together

**Comment by php-coder** [comment]
ci/openshift-jenkins/extended_conformance_gce fails with the following error:
```
TASK [gce-provision : Provision GCE resources] *********************************
Thursday 24 August 2017  14:27:09 +0000 (0:00:00.084)       0:00:02.686 ******* 
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/tmp/provision.sh"], "delta": "0:00:51.276171", "end": "2017-08-24 14:28:01.294004", "failed": true, "rc": 1, "start": "2017-08-24 14:27:10.017833", "stderr": "ERROR: (gcloud.compute.networks.create) Could not fetch resource:\n - Quota 'ROUTES' exceeded.  Limit: 300.0", "stderr_lines": ["ERROR: (gcloud.compute.networks.create) Could not fetch resource:", " - Quota 'ROUTES' exceeded.  Limit: 300.0"], "stdout": "", "stdout_lines": []}
```
ci/openshift-jenkins/extended_conformance_install_update fails with error:
```
• Failure [149.904 seconds]
[Conformance][networking][router] openshift router metrics
/go/src/github.com/openshift/origin/test/extended/router/metrics.go:192
  The HAProxy router
  /go/src/github.com/openshift/origin/test/extended/router/metrics.go:191
    should expose prometheus metrics for a route [It]
    /go/src/github.com/openshift/origin/test/extended/router/metrics.go:173

    Expected error:
        <*errors.errorString | 0xc4202d78e0>: {
            s: "timed out waiting for the condition",
        }
        timed out waiting for the condition
    not to have occurred

    /go/src/github.com/openshift/origin/test/extended/router/metrics.go:128
```

**Comment by php-coder** [comment]
@deads2k @enj PTAL

File: pkg/security/securitycontextconstraints/provider.go (hunk 9 above)
Comment (php-coder) [reply pkg/security/securitycontextconstraints/provider.go]: PR with the aforementioned changes has been submitted: https://github.com/openshift/origin/pull/16006

File: pkg/security/apis/security/v1/types.go [hunk 11]
```diff
@@ -50,6 +50,10 @@ type SecurityContextConstraints struct {
 	// of a VolumeSource (azureFile, configMap, emptyDir).  To allow all volumes you may use "*".
 	// To allow no volumes, set to ["none"].
 	Volumes []FSType `json:"volumes" protobuf:"bytes,8,rep,name=volumes,casttype=FSType"`
+	// AllowedFlexVolumes is a whitelist of allowed Flexvolumes.  Empty or nil indicates that all
+	// Flexvolumes may be used.  This parameter is effective only when the usage of the Flexvolumes
+	// is allowed in the "Volumes" field.
+	AllowedFlexVolumes []AllowedFlexVolume `json:"allowedFlexVolumes" protobuf:"bytes,21,rep,name=allowedFlexVolumes"`
```
Comment (enj) [inline pkg/security/apis/security/v1/types.go]: Is it OK for protobuf annotations to be out of order?

Comment (php-coder) [reply pkg/security/apis/security/v1/types.go]: If I correctly understood [this comment](https://github.com/openshift/origin/pull/15558#discussion_r130333789) then yes, it's OK.

**Comment by enj** [comment]
/retest
/lgtm
@php-coder please make sure the followups that Jordan wanted are taken care of.

**Comment by enj** [comment]
@pweil- please approve if you are good.

**Comment by php-coder** [comment]
Rebased.

**Comment by php-coder** [comment]
Rebased. PTAL.

**Comment by php-coder** [comment]
Test flake https://github.com/openshift/origin/issues/15900
/test cmd

---

```
=== RUN   TestRouterDuplications
--- FAIL: TestRouterDuplications (0.00s)
	router_test.go:752: Unable to start http server: listen tcp 172.18.9.82:8080: bind: address already in use
```
/test end_to_end

---

```
PLAY RECAP *********************************************************************
 [WARNING]: Could not create retry file '/usr/share/ansible/openshift-
ansible/playbooks/byo/config.retry'.         [Errno 13] Permission denied:
u'/usr/share/ansible/openshift-ansible/playbooks/byo/config.retry'
localhost                  : ok=91   changed=6    unreachable=0    failed=1   


Failure summary:

  1. Host:     localhost
     Play:     Verify Requirements
     Task:     openshift_health_check
     Message:  One or more checks failed
     Details:  check "package_version":
               Not all of the required packages are available at their requested version
               docker:1.12 
               Please check your subscriptions and enabled repositories.
```
/test extended_conformance_install_update

**Comment by adelton** [comment]
For testing this pull request, I try to create SCC with

```
volumes:
- configMap
- downwardAPI
- emptyDir
- persistentVolumeClaim
- projected
- secret
- flexVolume
allowedFlexVolumes:
- jezek
```

but I get

```
$ oc create -f restricted-flex-jezek.yaml
Error from server (BadRequest): error when creating "restricted-flex-jezek.yaml": SecurityContextConstraints in version "v1" cannot be handled as a SecurityContextConstraints: only encoded map or array can be decoded into a struct
```

What is the correct way to specify the list of drivers?

**Comment by php-coder** [comment]
> What is the correct way to specify the list of drivers?

@adelton Try this one:
```yaml
allowedFlexVolumes:
- driver: jezek
```

**Comment by adelton** [comment]
> Try this one:
> 
> ```
> allowedFlexVolumes:
> - driver: jezek
> ```

Thanks. Would it make sense to amend the documentation to show this syntax?

**Comment by php-coder** [comment]
> Would it make sense to amend the documentation to show this syntax?

The Trello card has `documentation` label, so this new feature will be documented. I'm going to do it after this PR will be merged. Nevertheless you can prepare PR for that ;)

**Comment by adelton** [comment]
From functional perspective, this feature works.

My only concern is an extremely long and not really helpful error message when I try to create pod with flexVolume and the driver is not listed in `allowedFlexVolumes`:

```
Error from server (Forbidden): error when creating "openshift-origin/test-flex-pod.json": pods "test-flex" is forbidden: unable to validate against any security context constraint: [spec.volumes[0]: Invalid value: "flexVolume": flexVolume volumes are not allowed to be used provider restricted-flex-jezek: .spec.securityContext.volumes[0].driver: Invalid value: "example.test/bind-mount": Flexvolume driver is not allowed to be used spec.volumes[0]: Invalid value: "flexVolume": flexVolume volumes are not allowed to be used spec.volumes[0]: Invalid value: "flexVolume": flexVolume volumes are not allowed to be used spec.volumes[0]: Invalid value: "flexVolume": flexVolume volumes are not allowed to be used spec.volumes[0]: Invalid value: "flexVolume": flexVolume volumes are not allowed to be used]
```

**Comment by liggitt** [comment]
> My only concern is an extremely long and not really helpful error message when I try to create pod with flexVolume and the driver is not listed in allowedFlexVolumes:

That's a pre-existing issue with the way SCC admission presents failures. A separate issue to improve that is probably a good idea.

**Comment by php-coder** [comment]
> That's a pre-existing issue with the way SCC admission presents failures. A separate issue to improve that is probably a good idea.

It looks like a regression or wrong merge. I'll look at it today more closely.

**Comment by php-coder** [comment]
> My only concern is an extremely long and not really helpful error message

@adelton I couldn't reproduce it. In my case the error is the following:
> Error from server (Forbidden): error when creating "flex-pod.yaml": pods "flexpod" is forbidden: unable to validate against any security context constraint: [provider restricted: .spec.securityContext.volumes[0].driver: Invalid value: "example-one.com/test": Flexvolume driver is not allowed to be used]

**Comment by php-coder** [comment]
End-to-end tests fails because of https://github.com/openshift/origin/issues/16144

**Comment by php-coder** [comment]
@deads2k @enj PTAL

**Comment by deads2k** [comment]
/approve 

@enj you own merge.

**Comment by enj** [comment]
/lgtm

No logic changes since last LGTM.

**Comment by adelton** [comment]
> I couldn't reproduce it. In my case the error is the following:

If you create bunch of differently named copies of `restricted` SCC, with

```
  groups:
  - system:authenticated
```

`developer` will have access to all of them even if they will not give the access to the flexVolume driver. And then the error message will repeat for all of them, but without showing the name of the SCC the error refers to it is really cryptic.

**Comment by php-coder** [comment]
> If you create bunch of differently named copies of restricted SCC,

Aha, I see. This is indeed a different case. Please, create a separate issue for that. Thanks!
